javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -0,0 +1,1729 @@
+      } else {
+        // To strip trailing zeros until the preferred scale is reached
+        val lastPow = BigTenPows.length - 1
+
+        @inline
+        @tailrec
+        def loop(i: Int, iq: BigInteger, scale: Long): (BigInteger, Long) = {
+          if (!iq.testBit(0)) {
+            val quotAndRem = iq.divideAndRemainder(BigTenPows(i))
+            if ((quotAndRem(1).signum() == 0) && (scale - i >= diffScale))
+              loop(if (i < lastPow) i + 1 else i, quotAndRem(0), scale - i)
+            else if (i != 1)
+              loop(1, iq, scale)
+            else
+              (iq, scale)
+          } else {
+            (iq, scale)
+          }
+        }
+
+        loop(1, quotAndRem(0), newScale0)
+      }
+    }
+    // To perform rounding
+    new BigDecimal(integerQuot, safeLongToInt(newScale), mc)
+  }
+
+  def divideToIntegralValue(divisor: BigDecimal): BigDecimal = {
+    if (divisor.isZero)
+      throw new ArithmeticException("Division by zero")
+
+    var integralValue: BigInteger = null
+    val newScale: Long = this.Scale.toLong - divisor.Scale
+    var varScale: Long = 0
+    val lastPow = BigTenPows.length - 1
+    if ((divisor.approxPrecision() + newScale > this.approxPrecision() + 1L) || this.isZero) {
+      /* If the divisor's integer part is greater than this's integer part,
+       * the result must be zero with the appropriate scale */
+      integralValue = BigInteger.ZERO
+    } else if (newScale == 0) {
+      integralValue = getUnscaledValue.divide(divisor.getUnscaledValue)
+    } else if (newScale > 0) {
+      val powerOfTen = powerOf10(newScale)
+      integralValue = getUnscaledValue.divide(divisor.getUnscaledValue.multiply(powerOfTen))
+      integralValue = integralValue.multiply(powerOfTen)
+    } else {// (newScale < 0)
+      val powerOfTen = powerOf10(-newScale)
+      integralValue = getUnscaledValue.multiply(powerOfTen).divide(divisor.getUnscaledValue)
+
+      // To strip trailing zeros approximating to the preferred scale
+      def loop(i: Int): Unit = if (!integralValue.testBit(0)) {
+        val quotAndRem = integralValue.divideAndRemainder(BigTenPows(i))
+        if ((quotAndRem(1).signum() == 0) && (varScale - i >= newScale)) {
+          varScale -= i
+          integralValue = quotAndRem(0)
+          loop(if (i < lastPow) i + 1 else i)
+        } else if (i != 1) {
+          loop(1)
+        }
+      }
+
+      loop(1)
+    }
+
+    if (integralValue.signum() == 0) zeroScaledBy(varScale)
+    else new BigDecimal(integralValue, safeLongToInt(varScale))
+  }
+
+  def divideToIntegralValue(divisor: BigDecimal, mc: MathContext): BigDecimal = {
+    val mcPrecision = mc.precision
+    val diffPrecision = this.precision() - divisor.precision()
+    val lastPow = BigTenPows.length - 1
+    val diffScale = this.Scale.toLong - divisor.Scale
+    var newScale = diffScale
+    val quotPrecision = diffPrecision - diffScale + 1
+    var quotAndRem = new Array[BigInteger](2)
```

Comment:
Try and make this a `val`, even if you have to redeclare it twice or thrice under different names.